### ISSUE TITLE
startup error

### ISSUE THREAD

**Issue by bgruening** [federicomarini/docker-phyloseq#2]
```
Downloading https://github.com/joey711/shiny-phyloseq/archive/HEAD.tar.gz
Loading required package: shiny
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
Warning in install.packages("BiocManager") :
  'lib = "/usr/local/lib/R/site-library"' is not writable
Error in install.packages("BiocManager") : unable to install packages
Calls: <Anonymous> ... source -> withVisible -> eval -> eval -> install.packages
Execution halted
```

It seems a normal startup triggers an installation:

```
galaxy@sn06:~$ docker run -i -i -p 8080:3434 quay.io/federicomarini/shinyphyloseq:v0.1
Unable to find image 'quay.io/federicomarini/shinyphyloseq:v0.1' locally

v0.1: Pulling from federicomarini/shinyphyloseq
aece8493d397: Pull complete 
ad8dc6239f74: Pull complete 
bc9e3e9ff617: Pull complete 
f4f98b882960: Pull complete 
614246e4f3d1: Pull complete 
d25c6c25d21d: Pull complete 
b3fe5508a6e1: Pull complete 
9bf9b0a40f30: Pull complete 
bd4d9be2bd29: Pull complete 
a5097744389a: Pull complete 
cbbf1467b697: Pull complete 
73ea020c1bdc: Pull complete 
15c8d1e55607: Pull complete 
46115fdeac25: Pull complete 
Digest: sha256:bcf96e8d0498654e1ca10b4451f4e1120d1b414689a3bd0f13f7b71d1863b8e0
Status: Downloaded newer image for quay.io/federicomarini/shinyphyloseq:v0.1
Downloading https://github.com/joey711/shiny-phyloseq/archive/HEAD.tar.gz
Loading required package: shiny
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://packagemanager.posit.co/cran/__linux__/jammy/latest/src/contrib/BiocManager_1.30.22.tar.gz'
Content type 'binary/octet-stream' length 499020 bytes (487 KB)
==================================================
downloaded 487 KB

* installing *binary* package ‘BiocManager’ ...
* DONE (BiocManager)

The downloaded source packages are in
	‘/tmp/RtmpQM82YT/downloaded_packages’
Loading required package: BiocManager
phyloseq package version:
1.46.0
biomformat package version:
1.30.0
shiny package version:
1.8.0
shinythemes package version:
1.2.0
ggplot2 package version:
3.5.0
data.table package version:
1.15.2
networkD3 package version:
0.4
^Cgenefilter package version:
1.84.0
grid package version:
4.3.2
gridExtra package version:
2.3
markdown package version:
1.12
rmarkdown package version:
2.25
png package version:
0.1.8
RColorBrewer package version:
1.1.3
scales package version:
1.3.0

Listening on http://0.0.0.0:3838

**Comment by federicomarini** [comment]
Yep, that is "correct", i.e. that is how it is defined in the way the app is provided (so far).
It also exactly ran like this for me too.

A plan for the coming versions is to detach this behavior and handle this in the installation steps where the docker image is built.

Thing is: I wanted to see if it even works, and now it does. That was my step 0.
I have been working a bit already on the Dockerfile step, but by `shiny::runGithub` it will still run the installation round.

Alternative that I was trying to explore: Insert the code of the app, "from Github", and have slight edits to avoid installation of packages - which we would have anyway pre-installed.
Probably little more weight of maintenance on the container providers. 

And yes, this is exactly why I like more the apps in a package.
Let alone that all the import statements that a package would handle elegantly AND "like a good citizen", here go pretty wild as one just has to call a bunch of times `library(...)`, so this can lead to unfortunate overwriting of functions.

Most notable example would be `select()` from `dplyr` AND from `AnnotationDbi` - both quite commonly found in common workflows.

**Comment by bgruening** [comment]
Interesting, TIL.
The problem is that we do not expect this to happen during runtime and containers are lock-down in some setups - .g. our. I had the Galaxy integration ready 20min after you wrote the mail, but the deployment did not work, because the user we run the container in has no write permission to the R installation dir.

**Comment by federicomarini** [comment]
Yeah, this really depends on the (probably not too optimal, especially for our context) manner the "app code" is put together.
Such things should just not happen in CRAN/Bioc packages. Or FWIW, in "any" fully fledged package - where the DESCRIPTION calls the shots, and we would stick to that & pick all things up while building the container...

**Comment by paulzierep** [comment]
I guess the issue of not being able to write the R install dir leads to:
```
Downloading https://github.com/joey711/shiny-phyloseq/archive/HEAD.tar.gz
Loading required package: shiny
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
Warning in install.packages("BiocManager") :
  'lib = "/usr/local/lib/R/site-library"' is not writable
Error in install.packages("BiocManager") : unable to install packages
Calls: <Anonymous> ... source -> withVisible -> eval -> eval -> install.packages
Execution halted
```

**Comment by federicomarini** [comment]
Actually, once the packages are in, this should still work.
This should be handled in some form of efficient form by the extra Dockerfile step.

Then again: That this form of installing dependencies is... a bit less efficient than packaging things up, well, we have been experiencing that on our skin directly 😉 

@paulzierep I hope you can have some fun in testing this kind of prototype a bit!

**Comment by bgruening** [comment]
> Actually, once the packages are in, this should still work.

What does that mean. What means "in"? I think its a bit confusing that R tries to install the same packages during container start which already have been installed during container build.

**Comment by paulzierep** [comment]
thanks  @federicomarini yes I will have some fun :)

I quickly made a clone and removed the `# install.packages("BiocManager")` part: https://github.com/paulzierep/shiny-phyloseq/blob/eb485ed1a4db2ea5bc1057a6098eef1ee0708f1c/install.R#L13C1-L13C34

Then the containers start up without install 
```
phyloseq_app <- shiny::runGitHub(repo = "shiny-phyloseq",
                                 username = "paulzierep", 
                                 launch.browser = FALSE, 
                                 port = 3838, 
                                 host = "0.0.0.0")
```

The R specific question would be how to avoid the BiocManager installation, without removing it from the shiny-phyloseq requirements, since it is already installed in the container. 
But this fix should work for now.

**Comment by federicomarini** [comment]
> > Actually, once the packages are in, this should still work.
> 
> What does that mean. What means "in"? I think its a bit confusing that R tries to install the same packages during container start which already have been installed during container build.

Well, that behavior is simply defined in here: https://github.com/joey711/shiny-phyloseq/blob/master/global.R#L3

@paulzierep : the fact that the app calls this line and this calls for an installation to start, I think it is driven by the desire of having the dependencies there "as one should". The usual answer to this issue, from my perspective, is "make it a package".

As a thought: The container could in principle *avoid* using the `shiny::runGitHub` call, and having it pick some locally stored folder (ideally, the clone of that, with the modifications in place) via `shiny::runApp()`. That, of course, brings in some weight of maintenance on us, container-providers.

tl;dr: an excellent item for the guidelines we would like to put together 😉

**Comment by paulzierep** [comment]
Could we try that as a quick fix: https://github.com/federicomarini/docker-phyloseq/pull/3 and see if it runs on galaxy ?
With some help, I would still be happy to work on a more generic solution, i.e. making a package ... but that will take some time.

**Comment by paulzierep** [comment]
@federicomarini do you know if the app can access the `shiny_input` folder ? 
My idea would be to modify the app to load any file that galaxy stores in there: https://github.com/paulzierep/shiny-phyloseq/issues/1

**Comment by federicomarini** [comment]
I have literally no clue till now about how the app (this one specifically) does get the data fed in.
This very much depends on how this (and every other app) is developed.

The other examples we have via apps that can be "launched and wait idle" (uploading at runtime) or "launch on a specific (set of) object(s)" are a bit easier to implement.

The modification of such a large app itself is probably best left to the main developer for that. Of course we could suggest a way to do so and by doing so enhance even more the range of applicability...

**Comment by paulzierep** [comment]
I made a hack to the app and @bgruening helped to set up the wrapper and the IT is working now on EU: https://usegalaxy.eu/?tool_id=interactive_tool_phyloseq&version=latest I will write a BlogPost about this soon ! Thanks a lot @federicomarini !
